### PR TITLE
Make gdb work with snapshot in Partial Execution.

### PR DESCRIPTION
**Pull request by itewqq**
### Which kind of PR do you create?

- [x] This PR only contains minor fixes.
- [ ] This PR contains major feature update.
- [ ] This PR introduces a new function/api for Qiling Framework.

### Coding convention?

- [ ] The new code conforms to Qiling Framework naming convention.
- [ ] The imports are arranged properly.
- [x] Essential comments are added.
- [ ] The reference of the new code is pointed out.

### Extra tests?

- [x] No extra tests are needed for this PR.
- [ ] I have added enough tests for this PR.
- [ ] Tests will be added after some discussion and review.

### Changelog?

- [ ] This PR doesn't need to update Changelog.
- [x] Changelog will be updated after some proper review.
- [ ] Changelog has been updated in my PR.

### Target branch?

- [x] The target branch is dev branch.

### One last thing

- [x] I have read the [contribution guide](https://docs.qiling.io/en/latest/contribution/)

-----

### CODE REVIEW

File: qiling/debugger/gdb/gdb.py [hunk 1]
```diff
@@ -65,6 +65,11 @@ def __init__(self, ql, ip, port):
         else:
             self.entry_point = self.ql.os.entry_point
 
+        # Only part of the binary file is debugged.
+        if self.ql.entry_point is not None and self.ql.exit_point is not None:
+            self.entry_point=self.ql.entry_point
```
Comment (aquynh) [inline qiling/debugger/gdb/gdb.py:70]: please put a space before and after `=`  (so " = ", not "=").

Comment (itewqq) [reply qiling/debugger/gdb/gdb.py]: Oh, my bad. Fixed now.

**Comment by xwings** [comment]
Hi,

We still need a elf header?

**Comment by itewqq** [comment]
@xwings Ok, sorry I didn't explain my context clearly. 

In fact, if you take a look at my python script, you will find that I am actually performing the so-called "Partial Execution" simulation, as mentioned in the [document snapshot](https://docs.qiling.io/en/latest/snapshot/). 

I first let the program run, hook at the main address and take a snapshot. Then I resume the snapshot, but directly assign the pc pointer to the beginning of the function I want to simulate.

In this way, I can simply construct the call_state (registers, memory mapping, dynamic library loading, etc) of the function, and at the same time avoid the tedious operation (including various hardware conditions such as NVRAM, and other processes Interaction) from the entry point of the elf to the part I am really interested in. I think this is a very important reason why I use qiling instead of directly using qemu and other large-scale simulations. After finished the static analysis, I only need to simulate the functions of interest, and then write exploit scripts...save me a coffee time.

But when I did this, I found that qiling could not start gdbserver (actually I could only use hook and print to debug), so I dig into the source code of qiling and found that it only takes a little work to make this become possible.

**Comment by aquynh** [comment]
Interesting! Will be nice if you can contribute a simple sample to examples/ so others can learn from this case.

**Comment by itewqq** [comment]
A simple example ```examples/hello_x8664_linux_part_debug.py``` is added based on the existing sleep_hello program.

**Comment by cq674350529** [comment]
@itewqq Just to complement: it's possible to emulate part of the code in binary and enable gdb without snapshot, please see https://github.com/qilingframework/qiling/issues/593.

Indeed, if running with snapshot, the gdb won't work as you describe.

**Comment by itewqq** [comment]
> @itewqq Just to complement: it's possible to emulate part of the code in binary and enable gdb without snapshot, please see #593.
> 
> Indeed, if running with snapshot, the gdb won't work as you describe.

@cq674350529 Well, my snapshot was taken at the beginning of ```main```... So I did not encounter the problem in this issue？

Can you explain how to emulate part of the code in binary and enable gdb without snapshot? And why if running with snapshot, the gdb won't work as me describe, have you tried https://github.com/qilingframework/qiling/pull/768/commits/d3d696fcaa13c6ad10d67eb7b54aa468d92bf863 ?

**Comment by cq674350529** [comment]
@itewqq Sorry for getting you misunderstand. There is nothing wrong with your pull request, and it can solve the issue https://github.com/qilingframework/qiling/issues/767 you encoutered.

What I mean is that without your fix, there may be some issues when running gdb with snapshot.

> Can you explain how to emulate part of the code in binary and enable gdb without snapshot?

As mentioned in https://github.com/qilingframework/qiling/issues/593, what we need is to hook at the beginning of `main`, and hijack `PC` to address you want ro run start.

```python
def sample():
    def partial_run(ql, *args, **kwargs):
        ql.reg.arch_pc = <address_to_run>    # hijack pc

    ql = Qiling([bin_path], rootfs_path, console=True, verbose=QL_VERBOSE.DEBUG)
    ql.hook_address(partial_run, <begin_of_main>)   
    ql.debugger = "127.0.0.1:1234"

    ql.run()
```
 I usually do partial code emulation in this way.

**Comment by itewqq** [comment]
Wow, amazing! I didn’t expect to do this before, because my main reference is Partial Execution from https://docs.qiling.io/en/latest/snapshot/ 😆 Thank you very much!!!

**Comment by cq674350529** [comment]
@itewqq Yeah, using snapshot mechanism can achieve the same goal. But I prefer this way, for it's more simple and clear.

**Comment by xwings** [comment]
Cool. Nice work buddy.

Thanks.

**Pull request merged**